File: 904.水果成篮.c
/*
 * @lc app=leetcode.cn id=904 lang=c
 *
 * [904] 水果成篮
 */

// @lc code=start

//滑动窗口
int totalFruit(int *fruits, int fruitsSize)
{
    int res = 0;
    int left = 0;
    int right = 0;
    int subl = fruits[left];
    int subr = fruits[right];

    for (right = 0; right < fruitsSize; right++)
    {
        if (fruits[right] == subl || fruits[right] == subr)
        {
            res = fmax(right - left + 1, res);
        }
        else
        {
            subl = fruits[right - 1];
            left = right - 1;
            while (left >= 1 && fruits[left - 1] == subl)
            {
                left--;
            }
            subr = fruits[right];
            res = fmax(res, right - left + 1);
        }
    }
    return res;
}

// @lc code=end

File: 904.水果成篮.py
#
# @lc app=leetcode.cn id=904 lang=python3
#
# [904] 水果成篮
#

# @lc code=start

class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        left = 0
        right = 0
        subr = fruits[left]
        subl = fruits[right]
        res = 0

        for right in range(len(fruits)):
            if fruits[right] in [subr, subl]:
                res = max(res, right - left + 1)
            else:
                subl = fruits[right - 1]
                left = right - 1

                while left >= 1 and fruits[left - 1] == subl:
                    left = left - 1
                subr = fruits[right]
                res = max(res, right - left + 1)
        return res

# @lc code=end

